pennylane/wires.py

```diff
@@ -0,0 +1,369 @@
+
+    else:
+        raise WireError(
+            "Wires must be represented by a number or string; got {} of type {}.".format(
+                wires, type(wires)
+            )
+        )
+
+
+class Wires(Sequence):
+    """
+    A bookkeeping class for wires, which are ordered collections of unique objects. The i'th object
+    addresses the i'th quantum subsystem.
+
+    There is no conceptual difference between registers of multiple wires and single wires,
+    which are just wire registers of length one.
+
+    Indexing and slicing this sequence will return another ``Wires`` object.
+
+    Args:
+         wires (Any): If iterable, interpreted as an ordered collection of unique objects representing wires.
+            Any other input type is converted into an iterable of a single entry, and hence interpreted as a
+            single wire.
+    """
+
+    def __init__(self, wires):
+
+        self.wire_tuple = _process(wires)
+
+        # check that all wires are unique
+        if len(set(self.wire_tuple)) != len(self.wire_tuple):
+            raise WireError("Wires must be unique; got {}.".format(wires))
+
+    def __getitem__(self, idx):
+        """Method to support indexing. Returns a Wires object representing a register with a single wire."""
+        return Wires(self.wire_tuple[idx])
+
+    def __len__(self):
+        """Method to support ``len()``."""
+        return len(self.wire_tuple)
+
+    def __contains__(self, item):
+        """Method checking if Wires object contains an object."""
+        if isinstance(item, Wires):
+            # If all wires can be found in this object, return True
+            if all(wire in self.wire_tuple for wire in item.wire_tuple):
+                return True
+        else:
+            return item in self.wire_tuple
+
+    def __repr__(self):
+        """Method defining the string representation of this class."""
+        return "<Wires = {}>".format(list(self.wire_tuple))
+
+    def __eq__(self, other):
+        """Method to support the '==' operator. This will also implicitly define the '!=' operator."""
+        # The order is respected in comparison, so that ``assert Wires([0, 1]) != Wires([1,0])``
+        if isinstance(other, self.__class__):
+            return self.wire_tuple == other.wire_tuple
+        return False
+
+    def __hash__(self):
+        """Implements the hash function."""
+        return hash(repr(self.wire_tuple))
+
+    def toarray(self):
+        """Returns a numpy array representation of the Wires object.
+
+        Returns:
+            ndarray: array representing Wires object
+        """
+        return np.array(self.wire_tuple)
+
+    def tolist(self):
+        """Returns a list representation of the Wires object.
+
+        Returns:
+            List: list representing Wires object
+        """
+        return list(self.wire_tuple)
+
+    def index(self, wire):
+        """Overwrites a Sequence's ``index()`` function which returns the index of ``wire``.
```

Comment:
Does this method work as follows?
```python
>>> wires = Wires([2, 6, 1])
>>> wires.index(1)
3
```
Is the reason for the overwriting so that you can also do this (which isn't support by `Sequence`)?:

```python
>>> wires = Wires([2, 6, 1])
>>> wires.index(Wires(1))
3
```

Reply:
If so, nice

Reply:
No, that would be the ``indices`` function. The ``index`` function needs to be overwritten because it does not work when single wires are ``Wires`` objects...

Reply:
We need both methods because index() should return an int. It is awful to always check if it is a list instead.